bebraven/platform
# Implemented LinkedIn authorization flow

**Task**: https://app.asana.com/0/1174274412967132/1191740236188816

**Development environment breaking changes**
- You need to run a DB migration
- Add `LINKED_IN_CLIENT_{ID, SECRET}` to your `.env` file

**Test**:

- Go to `https://platformweb/linked_in/login` endpoint on desktop browsers (Chrome, Safari, Firefox)

<img width="1792" alt="Screen Shot 2020-09-14 at 7 01 12 PM" src="https://user-images.githubusercontent.com/62911934/93156398-e71ddd00-f6bc-11ea-8130-420c2ade1ce4.png">

- Using `linked_in/signin` on Heroku test app on mobile browser (Safari)
<img height="500" alt="IMG_1601" src="https://user-images.githubusercontent.com/62911934/93156577-3b28c180-f6bd-11ea-9c24-e3bac49fb8fb.PNG">

- Create a Project in Content Editor. Add Section, then iFrame, with URL `https://platformweb/linked_in/login`, save the Project, and add it as a Canvas assignment. 

<img width="1792" alt="Screen Shot 2020-09-15 at 2 31 55 PM" src="https://user-images.githubusercontent.com/62911934/93267322-5dbfe680-f760-11ea-8511-90218e20039c.png">

<img width="1792" alt="Screen Shot 2020-09-15 at 2 33 15 PM" src="https://user-images.githubusercontent.com/62911934/93267382-77612e00-f760-11ea-8ce9-0292adfb3fb2.png">

<img width="1792" alt="Screen Shot 2020-09-15 at 2 33 24 PM" src="https://user-images.githubusercontent.com/62911934/93267390-78925b00-f760-11ea-8f1f-d26c6e6c4a51.png">

- Once you've accepted using one the above methods, look up the access token in `devdb`
```
select linked_in_access_token from users where id=<your user ID>;`
```

- Query the API and make sure you get a JSON blob of data back:
```
curl 'https://api.linkedin.com/v2/me' \
-X GET \
-H "Authorization: Bearer <linked_in_access_token>"
```

- Query the API using `LinkedInConnection` in `devc`:
> require 'linked_in_connection'
> LinkedInConnection.get_request('/v2/me', Users.find(your_user_id).linked_in_access_token)

<img width="1792" alt="Screen Shot 2020-09-15 at 3 17 01 PM" src="https://user-images.githubusercontent.com/62911934/93271193-826b8c80-f767-11ea-83e7-8f936c5f6b88.png">

- You can also check your LinkedIn profile > Settings > Data Privacy > Other applications (https://www.linkedin.com/psettings/permitted-services) and see that Braven is listed. Note: to test this again with the same user, you'll need to "Remove" the Braven permissions. 

<img width="1792" alt="Screen Shot 2020-09-15 at 12 33 36 PM" src="https://user-images.githubusercontent.com/62911934/93256182-d2d6f000-f74f-11ea-9dac-3dd19f17c1b1.png">

**Details**

**Testing**
- Added `ci_exclude` tag for potentially flaky tests that talk to LinkedIn
- Updated `app.json` so Heroku CI skips `ci_exclude` tests
- Updated Testing section in `README`

**LinkedIn**
- We can't embed the LinkedIn authorization flow in our Canvas iframe (it violates the ancestor policy that LinkedIn has set)
- Need to whitelist redirect endpoint (`https://platformweb/linked_in/auth_redirect`) to Braven Accelerator app on LinkedIn (https://www.linkedin.com/developers/apps/5004876/settings)
- You'll also need to whitelist the Heroku Review app if you want to use that to test (just don't forget to delete it)
- The authorization code you get back from LinkedIn's redirect is valid for 30 minutes for getting the access token.

**New window**
- For window features, see: https://developer.mozilla.org/en-US/docs/Web/API/Window/open#Window_features
- To get a new window (instead of a new tab), you need to specify `width` and `height` in the Javascript when creating the new window. 
- To get the window to center before it renders, you need to specify `left` and `top` in the window features. 

**React examples**
- https://medium.com/hackernoon/using-a-react-16-portal-to-do-something-cool-2a2d627b0202
- https://dev.to/dinkydani21/how-we-use-a-popup-for-google-and-outlook-oauth-oci

**Future work**
~~Content Editor output: https://app.asana.com/0/1174274412967132/1193832365458406~~
- Prevent multiple pop-up windows from opening: https://app.asana.com/0/1174274412967132/1194517634372073
- More UI feedback: https://app.asana.com/0/1174274412967132/1193832365458401
